src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresAccessPolicyDAO.scala

```diff
@@ -745,22 +745,64 @@ class PostgresAccessPolicyDAO(protected val writeDbRef: DbReference, protected v
+        // then delete all the resource policies
+        // then delete all the groups from the first query
+        samsql"""delete $fromAndWhereFragment""".update().apply()
+
+        if (groupPKsToDelete.nonEmpty) {
+          samsql"""delete from ${GroupTable as g}
+                 where ${g.id} in ($groupPKsToDelete)""".update().apply()
+        }
+      }).attempt
+      _ <- handleResult(samRequestContext, writeResult, groupPKsToDelete)
+    } yield ()
+  }
+
+  private def handleResult(samRequestContext: SamRequestContext, writeResult: Either[Throwable, AnyVal], groupPKsToDelete: List[GroupPK]) = {
+    writeResult match {
+      case Left(fkViolation: PSQLException) if fkViolation.getSQLState == PSQLStateExtensions.FOREIGN_KEY_VIOLATION =>
+        handleForeignKeyViolation(samRequestContext, groupPKsToDelete)
+      case Left(e) => IO.raiseError(e)
+      case Right(_) => IO.unit
+    }
+  }
 
-      val groupPKsToDelete = samsql"""select ${p.result.groupId} $fromAndWhereFragment""".map(_.get[GroupPK](p.resultName.groupId)).list().apply()
+  private def handleForeignKeyViolation(samRequestContext: SamRequestContext, groupPKsToDelete: List[GroupPK]) = {
+    for {
+      problematicGroups <- getGroupsCausingForeignKeyViolation(samRequestContext, groupPKsToDelete)
+      _ <- IO.raiseError[Unit](new WorkbenchExceptionWithErrorReport( // throws a 500 since that's the current behavior
+        ErrorReport(StatusCodes.InternalServerError, s"Foreign Key Violation while deleting groups: ${problematicGroups}")))
+    } yield ()
+  }
 
-      samsql"""delete $fromAndWhereFragment""".update().apply()
+  private def getGroupsCausingForeignKeyViolation(samRequestContext: SamRequestContext, groupPKsToDelete: List[GroupPK]) = {
+    val g = GroupTable.syntax("g")
+    val pg = GroupTable.syntax("pg") // problematic group
+    val gm = GroupMemberTable.syntax("gm")
 
-      if (groupPKsToDelete.nonEmpty) {
-        samsql"""delete from ${GroupTable as g}
-                 where ${g.id} in ($groupPKsToDelete)""".update().apply()
-      }
-    })
+    readOnlyTransaction("getGroupsCausingForeignKeyViolation", samRequestContext) { implicit session =>
+      samsql"""select ${g.result.id}, ${g.result.name}, ${pg.result.name}
+                     from ${GroupTable as g}
+                     join ${GroupMemberTable as gm} on ${g.id} = ${gm.memberGroupId}
+                     join ${GroupTable as pg} on ${gm.groupId} = ${pg.id}
+                     where ${g.id} in
+                         (select distinct ${gm.result.memberGroupId}
+                          from ${GroupMemberTable as gm}
+                          where ${gm.memberGroupId} in ($groupPKsToDelete))"""
+        .map(rs =>
+          Map("groupId" -> rs.get[GroupPK](g.resultName.id).value.toString,
+            "groupName" -> rs.get[String](g.resultName.name),
+            "still used in group:" -> rs.get[String](pg.resultName.name)))
```

Comment:
might be more readable if these were collected into a single list so the output looked something like 
`groupId -> 12, groupName -> 'example', still used in group: -> [34, 45, 56]` , but not blocking

Reply:
updated:
```
{"causes":[],"message":"Foreign Key Violation(s) while deleting group(s): List(Map(groupId -> 7238732, groupName -> billing-project_aj-ca-1206_owner, still used in group(s): -> {workspace_1b86b9cd-35f2-4867-ad53-54be57a65856_project-owner,workspace_77fa1cce-b004-45c9-8429-da403cf905eb_project-owner,workspace_9845724b-34d9-40ca-9452-d391677ff277_project-owner}))","source":"sam","stackTrace":[],"statusCode":500}%
```